### PR TITLE
Add `ActiveSupport::Testing::NotificationAssertions` test helper module

### PR DESCRIPTION
**Pull request by larouxn**
### Motivation / Background

This Pull Request has been created because it's currently cumbersome to test that a certain code block/action triggers an `ActiveSupport::Notifications::Event` to be emitted. It would be ideal to have some helpers to `assert` against such event emission. Have heard such helpers could be helpful internal to Rails as well.

Notes
- Planning to clean up various parts of `rails/rails` with these helpers as a follow-up. ([WIP branch](https://github.com/larouxn/rails/tree/migrate_everywhere_to_event_assertions_helpers))
- Already have ideas for ways we can make this module even more useful: [comment from below](https://github.com/rails/rails/pull/53065#issuecomment-2452017030).

Co-authored by @yishus and @seanpdoyle :pray: 

### Detail

This Pull Request adds a new test helpers module called `ActiveSupport::Testing::NotificationAssertions`.

### Additional information

- An earlier form of these test helpers has been in use in production CI for our Rails application for over a month.
- Attempted to include method level documentation with a similar style to nearby code.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by larouxn** [comment]
[CI failure](https://buildkite.com/rails/rails/builds/111874#019232b6-6511-48b8-90da-dad3588c2266) is `ActionCable` specific and thus unlikely related to this PR's changes. Could use a retry.

```console
27 09 2024 09:12:02.418:ERROR [launcher.sauce]: Can not start chrome 70
27 09 2024 09:12:02.420:ERROR [karma-server]: UnhandledRejection
```

File: activesupport/test/notifications/event_assertions_test.rb
Comment (seanpdoyle) [inline activesupport/test/notifications/event_assertions_test.rb:23]: Could the assertions in this style utilize the [assert_raises](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Testing/Assertions.html#method-i-assert_raises) `:match` option?

```suggestion
        assert_raises Minitest::Assertion, match: /No post.submitted events were found/ do
          assert_event("post.submitted", title: "Cool Post") { nil } # no events
        end
```

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:37]: Other test helpers in this style (like [ActionMailer::TestHelper](https://edgeapi.rubyonrails.org/classes/ActionMailer/TestHelper.html#method-i-capture_emails) and [ActionCable::TestHelper](https://edgeapi.rubyonrails.org/classes/ActionCable/TestHelper.html#method-i-capture_broadcasts)) expose a `capture_`-prefixed helper that the assertions also invoke. 

Have you explored what that might involve for events? Maybe something like:

```suggestion
      def assert_event(name, payload = nil, &block)
        events = capture_events(name, &block)
        selected_events = events.select { |event| event.name == name }
        assert_not_empty(selected_events, "No #{name} events were found.")

        return if payload.nil?

        event = selected_events.find { |event| event.payload == payload }
        assert_not_nil(
          event,
          "No #{name} event with payload #{payload} was found. " \
            "Did you mean one of these payloads: #{selected_events.map(&:payload).join("\n")}?",
        )
      end
      
      def capture_events(name, &block)
        events = []

        ActiveSupport::Notifications.subscribe(name) do |*args|
          events << ActiveSupport::Notifications::Event.new(*args)
        end

        yield
        
        events
      end
```

File: activesupport/test/notifications/event_assertions_test.rb
Comment (larouxn) [reply activesupport/test/notifications/event_assertions_test.rb:23]: Great suggestion. Wasn't aware of that `assert_raises` parameter. Nice! Will push a cleanup.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb:37]: Had not noticed that pattern before. Will be quite nice to de-duplicate the event subscribing and yielding pattern with a simple, reusable `capture_events` private method. Will push a cleanup.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:66]: What makes this better as a `private` method? The Mailer and Cable test helpers expose their `capture_`-prefixed methods as public. That could add flexibility for tests that need more specificity between asserting a single event and a count of total events.

```suggestion
        def capture_events(name, &block)
```

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:71]: If you omit the `*` from the argument list and declare a block with an arity of 1, the parameter will be an already constructed instance of `ActiveSupport::Notifications::Event`:

```suggestion
          ActiveSupport::Notifications.subscribe(name) { |event| events << event }
```

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:17]: `event` is a fairly generic term. Since this is an Active Support test assertion, it's possible that it could collide with application-side domain terminology.

Have you considered other terminology from the domain of the [ActiveSupport::Notifications](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html) module that's more specific? Instances of `Event` are made available to the blocks, but those events collect instrumentation data that's being published to subscribers.

I'm struggling to suggest a specific improvement, but I wonder if there are alternatives that are worthwhile to consider. Maybe even `assert_notification_event` or `assert_instrumentation_event` could work.

Comment (seanpdoyle) [reply activesupport/lib/active_support/notifications/event_assertions.rb:17]: Have you explored existing patterns of use throughout the rest of the `rails/rails` test suites? There are various calls throughout that could utilize the assertions you're introducing in this PR. After a cursory search, I found:

* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actioncable/test/channel/base_test.rb#L201-L263
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actioncable/test/server/broadcasting_test.rb#L15-L53
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionmailbox/test/unit/mailbox/notifications_test.rb#L9-L31
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionmailer/test/base_test.rb#L966-L992

Some tests even utilize the [ActiveSupport::Notifications.subscribed](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-subscribed) variety of call.

* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionpack/test/dispatch/routing/instrumentation_test.rb#L27-L32
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionpack/test/dispatch/middleware_stack_test.rb#L195-L213

For what it's worth, my anecdotal findings make me think that the favored terminology for the concept is [instrumentation](https://github.com/search?q=repo%3Arails%2Frails%20instrumentation_test.rb&type=code), so it could be worthwhile to incorporate that into the name of the assertion methods or module.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb]: Hadn't considered the flexibility use case. It's a good call out. Don't need to make it private.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb]: > `event` is a fairly generic term. Since this is an Active Support test assertion, it's possible that it could collide with application-side domain terminology.

Completely agree. Been on-and-off thinking about what a more specific name might look like. Considering the Rails API docs states the following:

> `ActiveSupport::Notifications` provides an instrumentation API for Ruby.

I do agree that `assert_instrumentation_event` seems pretty accurate and would have a pretty low chance of name collision as well.

> Have you explored existing patterns of use throughout the rest of the `rails/rails` test suites? There are various calls throughout that could utilize the assertions you're introducing in this PR.

That cleanup opportunity was mentioned when I first proposed upstreaming this module. It's definitely something I'd like to do, a sweep over the codebase. Though I will probably do so in a follow-up PR. May look into opening one soon, stacked on this, as the implementation in here is slowly but surely maturing.

Much appreciate the cursory search. :pray:

Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb]: Briefly looked into updating the existing `ActiveSupport::Notifications` usage in tests. It looks like we should probably add an `unsubscribe` in these helpers as that's done in said tests. Should ideally ensure no side effects from using these helpers as well.

Have pushed a change to add `unsubscribe` the shared `capture_events` method.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:76]: The [ActiveSupport::Notifications.subscribed](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html#module-ActiveSupport::Notifications-label-Subscribe+While+a+Block+Runs) variation will unsubscribe after the passed block is invoked.

This could be simplified:

```suggestion
      def capture_instrumentation_events(...)
        events = []
        ActiveSupport::Notifications.subscribed(->(e) { events << e }, ...)
        events
      end
```

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:1]: What do you think about renaming the module to `ActiveSupport::Notifications::TestHelper` and the file to `activesupport/lib/active_support/notifications/test_helper.rb`? 

Other framework-specific test modules are named similarly:

* https://github.com/rails/rails/blob/main/activejob/lib/active_job/test_helper.rb
* https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/test_helper.rb
* https://github.com/rails/rails/blob/main/actionmailbox/lib/action_mailbox/test_helper.rb

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:66]: Since this method is public, it should have a method-level comment block to serve as documentation. It would be useful to add test coverage as well.

Comment (seanpdoyle) [reply activesupport/lib/active_support/notifications/event_assertions.rb:66]: Also, what do you think about renaming it to `capture_instrumentation_events` to more closely match the other new methods?

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb:1]: I think that's fine. For context, I named it `EventAssertions` after the `QueryAssertions`, though I do notice now that such is within a `testing` directory. Thus, maybe `TestHelper` makes sense for this PR's addition. 

- https://github.com/rails/rails/blob/main/activerecord/lib/active_record/testing/query_assertions.rb

Comment (seanpdoyle) [reply activesupport/lib/active_support/notifications/event_assertions.rb:1]: [ActiveRecord::Testing::QueryAssertions](https://github.com/rails/rails/blob/v7.2.1/activerecord/lib/active_record/testing/query_assertions.rb) is a good counter example. With that in mind, I do think `EventAssertions` could work. What do you think about moving it to the `testing/` directory?

Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb:1]: I think moving to `testing` directory makes sense given the use case of `QueryAssertions`, very similar to this module. Naming wise, I worry `ActiveSupport::Testing::Assertions` is a bit too generic. What about `ActiveSupport::Testing::InstrumentationAssertions`?

In fact, I just checked `ActiveSupport::Testing::Assertions` already exists as the primary Rails minitest enhancement/override method declaration file. As not everyone will need these event assertion methods I think a distinct name like `InstrumentationAssertions` might be best.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/notifications/event_assertions.rb:77]: I'm sorry to have missed this before, but the `name` argument isn't being used in the current implementation.

The [ActiveSupport::Notifications#subscribed](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-subscribed) method accepts an optional `pattern` argument, which filters events by String or Regexp name.

Since `capture_instrumentation_events` can be used independent from the assertions, what do you think about:

* renaming `name` to `pattern` (to mirror the underlying method's argument name
* making `pattern` optional
* passing `pattern` to the `subscribed` call

```suggestion
      def capture_instrumentation_events(pattern = nil, &block)
        events = []
        ActiveSupport::Notifications.subscribed(->(e) { events << e }, pattern, &block)
        events
      end
```

That change has the potential to simplify the rest of the implementations, especially the ones that utilize `selected_events = events.select { |event| event.name == name }`.

Since the `capture_instrumentation_events` can handle the filtering, any object in the returned `events` array will be guaranteed to have the expected name.

Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb:77]: No worries, so I didn't notice either. Been some long days recently. :sweat_smile: 

I think renaming to `pattern` in `capture_instrumentation_events` makes sense, especially since this method is much closer to the actual event subscription and all, makes sense to shadow the underlying parameter. The built in filtering is also ideal cleanup wise.

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb:1]: Ah, I misunderstood your suggestion. Have renamed to `InstrumentationAssertions` and moved under the `ActiveSupport::Testing` namespace and directory.

As for including the module or not, I think it makes sense to include in `TestHelp` for best user experience. Though I don't think we need to include it in these other places that `QueryAssertions` is being included in.

image

File: activesupport/CHANGELOG.md
Comment (larouxn) [inline activesupport/CHANGELOG.md:23]: Given the extremely helpful review that's greatly shaped this contribution, I think it makes sense to include you hear as well @seanpdoyle. :pray:

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (seanpdoyle) [inline activesupport/lib/active_support/testing/instrumentation_assertions.rb:73]: Does this work with `&block` as a block argument?

```suggestion
        ActiveSupport::Notifications.subscribed(->(e) { events << e }, pattern, &block)
```

File: activesupport/lib/active_support/notifications/event_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/notifications/event_assertions.rb]: Just had the thought, suppose the other methods `name` param and docs should probably be updated to `pattern` as well so peeps know they can use such. Otherwise peeps likely think only string is supported.

File: railties/lib/rails/test_help.rb
Comment (seanpdoyle) [inline railties/lib/rails/test_help.rb:22]: Since this is nested within an `defined?(ActiveRecord::Base)`, it will only work when Active Record is available.

Active Support can be used independently from other parts of the framework.

If making the module globally available is the goal, this should be moved to [ActiveSupport::TestCase](https://github.com/rails/rails/blob/fa152fba87f3a11790783207e6e84e58d345c012/activesupport/lib/active_support/test_case.rb#L144-L153).

If it's opt-in, it should be required in [lib/active_support/test_case.rb](https://github.com/rails/rails/blob/fa152fba87f3a11790783207e6e84e58d345c012/activesupport/lib/active_support/test_case.rb#L3-L18) and included ad-hoc test-by test in the application code.

I can't really weigh in on which it should be, I defer that to Core.

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/testing/instrumentation_assertions.rb:73]: It does indeed. Nice :raised_hands:

File: railties/lib/rails/test_help.rb
Comment (larouxn) [reply railties/lib/rails/test_help.rb:22]: Good eye. Moving to `ActiveSupport::TestCase`. Let's see what folks say.

Comment (larouxn) [reply railties/lib/rails/test_help.rb:22]: Good eye. Will move to `ActiveSupport::TestCase`. Let's see what folks say.

**Comment by larouxn** [comment]
For Rails core, the `railties` label can be removed. This PR no longer touches that component. :pray:

As for the [CI failure](https://buildkite.com/rails/rails/builds/112158#019252c0-bfe7-4dd4-b569-86e9a74d50f4), it appears to be a flaky/busted test. Found it [failing on another PR](https://buildkite.com/rails/rails/builds/112148#019251e0-5e38-4e62-bf9b-d65087bed864). Oddly it passes locally on `main` and this branch.

```console
Failure:
Rails::Command::RoutesTest#test_rails_routes_with_expanded_option [test/commands/routes_test.rb:249]:
--- expected
+++ actual
@@ -3,7 +3,7 @@
 Verb              | 
 URI               | /assets
 Controller#Action | Propshaft::Server
-Source Location   | propshaft (1.0.0) lib/propshaft/railtie.rb:43
+Source Location   | propshaft (1.1.0) lib/propshaft/railtie.rb:47
 --[ Route 2 ]--------------
 Prefix            | cart
 Verb              | GET
```

Will rebase on `main` in the morning, currently up to date, and see if that helps assuming it'll be fixed by then.

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (davidstosik) [inline activesupport/lib/active_support/testing/instrumentation_assertions.rb:28]: I got confused and mistook `payload` for `pattern`.
I guess scenarios in which there will be a lot of events on the right pattern but mismatched payload are gonna be super rare, so canceling my previous comment.

 
<details><summary>(I'll leave the content here if you're curious, but please ignore.)</summary>

Could this list become way too big to be reasonably read?
Could joining with `\n` cause a _scroll bomb_?

I was going to suggest the use of `DidYouMean`, but it doesn't look like the Rails codebase is using it much.
<details><summary>For the record, this is what it could look like.</summary>

```rb
error_message = "No #{pattern} event with payload #{payload} was found."
corrections = DidYouMean::SpellChecker.new(dictionary: events.map(&:payload)).correct(pattern)
if corrections.any?
  error_message += " Did you mean one of these payloads: #{corrections.join(", ")}?"
end
assert_not_nil(event, corrections)
```

</details> 

</details> 

Is it actually useful to display other payloads in the error message?

I can imagine multiple payload hashes with multiple keys in each of them could disrupt the flow of error messages, and am unsure this is actually the level of error messaging we need.

A developer could instead easily add a `capture_instrumentation_events` and debug to see what they're missing.

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (davidstosik) [inline activesupport/lib/active_support/testing/instrumentation_assertions.rb:60]: Very nitpicky but I find the line a bit long, hard to read in a typical editor, so I'd suggest ditching the one-liner:
```suggestion
        error_message = if pattern
          "Expected no events for #{pattern} but found #{events.size}"
        else
          "Expected no events but found #{events.size}"
        end
```

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (davidstosik) [inline activesupport/lib/active_support/testing/instrumentation_assertions.rb:28]: Should there be a line return before the first suggested payload?
```suggestion
            "Did you mean one of these payloads:\n#{events.map(&:payload).join("\n")}?",
```

File: activesupport/test/testing/instrumentation_assertions_test.rb
Comment (davidstosik) [inline activesupport/test/testing/instrumentation_assertions_test.rb:30]: (nitpick)
```suggestion
        assert_raises(Minitest::Assertion, match: /No post.submitted event with payload {:title=>"Cool Post"} was found. Did you mean one of these payloads: {:title=>"Cooler Post"}\n{:title=>"Coolest Post"}\?/) {
```

File: activesupport/test/testing/instrumentation_assertions_test.rb
Comment (davidstosik) [inline activesupport/test/testing/instrumentation_assertions_test.rb:80]: Another nitpick, but this would match too:
```suggestion
        events = capture_instrumentation_events(/post\./) do # regexp pattern
```

**Review by davidstosik — COMMENTED** [review]
Looks great to me! 👍🏻 

Another general nitpick: is it standard Rails behaviour to not add periods at the end of error messages?

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/testing/instrumentation_assertions.rb:28]: Fair points all around, especially about having `capture_instrumentation_events` available for debugging. Originally wrote this assertion helper with this `payload` error message before `capture_instrumentation_events` existed. I think now since that's around we can simplify this whole thing.

**Comment by larouxn** [comment]
> is it standard Rails behaviour to not add periods at the end of error messages?

Hmm, good question. Looking at a few places such as `QueryAssertions` and `ActiveSupport::Testing::Assertions` it does appear that error messages usually _don't_ end with periods. Have cleaned up this new module's accordingly.

https://github.com/rails/rails/blob/797790828bae932c41a353b30f799c179966840c/activerecord/lib/active_record/testing/query_assertions.rb#L25-L29

https://github.com/rails/rails/blob/797790828bae932c41a353b30f799c179966840c/activesupport/lib/active_support/testing/assertions.rb#L21-L24

File: activesupport/CHANGELOG.md
Comment (eileencodes) [reply activesupport/CHANGELOG.md:23]: Changelog entries need to be at the top of the file.

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (eileencodes) [inline activesupport/lib/active_support/testing/instrumentation_assertions.rb:18]: I've looked at this PR a few times and I keep getting hung up `instrumentation` being in the class and assertion names. It feels like a mouthful and prone to misspellings. I think something like `assert_event_payload` would be better. I'm curious what other core team members think about the name and if they have other ideas I'm not thinking of.

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/testing/instrumentation_assertions.rb:18]: Indeed, `Instrumentation` is bit complex naming wise. In fact, the original naming schema was all based around just `event` such as `assert_event` and `assert_no_events`. We pivoted to the current naming scheme as we thought the former was a bit too generic both semantically and such that it could cause naming conflicts.

As for `Instrumentation`, that was chosen as that's how these events are referred to in the documentation:

> `ActiveSupport::Notifications` provides an instrumentation API for Ruby.

That said, we could revert to `event` alone based naming or perhaps `assert_notification_event` or otherwise? Open to suggestions of course.

Comment (larouxn) [reply activesupport/lib/active_support/testing/instrumentation_assertions.rb:18]: Ideas from me

- `assert_event` (original)
  - pros: concise, clean
  - bad: low context, naming collision prone
- `assert_instrumentation_event` (current)
  - pros: high context, low chance of collision
  - cons: long, harder to say/spell, perhaps too verbose
- `assert_notifications_event`
  - pros: relatively high context, less hard to spell/say than `instrumentation`
  - cons: long, could use `notification` I suppose
- `assert_instrumented_event` or `assert_subscribed_event`
  - pros: relatively high context
  - cons: semantically feels like it should have `emitted` at the end or something similar

as well as from Matthew D

> IMO: Instrument is the verb; what it produces is notifications
> `assert_notification` / `assert_no_notifications`

and Christian Bica

> As `ActiveSupport::Notifications` emits events I'd use `assert_emits_event`

File: activesupport/test/testing/notification_assertions_test.rb
Comment (larouxn) [inline activesupport/test/testing/notification_assertions_test.rb:33]: Added this Ruby 3.4 conditional support in response to the following new CI failure.
https://buildkite.com/rails/rails/builds/112693#0192948f-8262-43f3-ba5c-bec1da067b09

File: activesupport/lib/active_support/testing/instrumentation_assertions.rb
Comment (larouxn) [reply activesupport/lib/active_support/testing/instrumentation_assertions.rb]: :wave: @eileencodes, when you have a moment, please let me know what you think of the new naming scheme and any other feedback. I think it's in a pretty good state now and would like to get it merged for 8.0 if possible. :pray: 

Can also use this new module to cleanup much of the [existing rails/rails codebase](https://github.com/rails/rails/pull/53065#discussion_r1780342475) once merged. :slightly_smiling_face:

File: activesupport/test/testing/notification_assertions_test.rb
Comment (davidstosik) [reply activesupport/test/testing/notification_assertions_test.rb:33]: To avoid the `RUBY_VERSION` conditional, you could also do this:

```rb
        match = /No post.submitted notification with payload #{{title: "Cool Post"}} was found/
```
The string interpolation would rely on the current Ruby version's implementation of `Hash#to_s`.

Though at this point, one could wonder is that writing a test or duplicate the app logic? 😅

Comment (larouxn) [reply activesupport/test/testing/notification_assertions_test.rb:33]: Ah that's true, could just `to_s` the Hash right in the test to get the relevant implementation for whichever Ruby is running. _Is_ reimplementing the actual code to a degree so I generally prefer the current conditional approach though the suggestion is a bit cleaner.

For context, when I ran into the Ruby 3.4 preview 2 CI failure I searched around the Rails codebase and found various other places using Ruby version conditionals. One linked below.

https://github.com/rails/rails/blob/main/actionpack%2Ftest%2Fdispatch%2Fexception_wrapper_test.rb#L115-L119

**Comment by larouxn** [comment]
Currently failing CI due to errors running `test:integration:sidekiq`. 🤔 ([logs](https://buildkite.com/rails/rails/builds/113138#0192d1ce-552c-425c-8aea-1c7f93918a35))

**Comment by larouxn** [comment]
From my time working on migrating the existing Rails test cases to use these helpers ([WIP branch](https://github.com/larouxn/rails/tree/migrate_everywhere_to_event_assertions_helpers)), a few potential improvements have arisen. They don't necessarily need to happen, let alone in v1, but I think such will make the helpers more useful. Please see the ideas along with rough draft pseudo code grade examples below and let me know what y'all think. CC @seanpdoyle, @yishus

### Payload partial match assertion

Helps utilize `assert_notification` in cases with extraneous fields.

```ruby
assert_notification("post.create", { name: "great post" }, partial_match: true) do
  Post.create(name: "great post") # emits notification with "great post" name as well as metadata fields we don't care about
end
```

### Multiple event assertion

Helps prevent needing to stack `assert_notification` calls thanks to new plural `assert_notifications`.

```ruby
assert_notifications("post.create", [{ name: "great post" }, { name: "best post" }]) do
  Post.create(name: "great post")
  Post.create(name: "best post")
end
```

### REGEX/match assertion

Helps when we want to assert some subset of a value.

```ruby
assert_notification("post.create", { name: /COOL/ }) do
  Post.create(name: "COOL POST AMAZING WOW WOW WOW")
end
```

### Presence assertion

Helps when we don't care about the value, just that the key is present with something other than `nil`. Can also be used to assert that a key is set to `nil` or perhaps key is not present.


```ruby
assert_notification("post.create", { name: :exists, body: nil, tag: :missing }) do
  Post.create(name: "wow", body: nil) # matches name is present, body is nil, and tag is not present
end
```

**Review by eileencodes — APPROVED** [review]
I like the naming a lot better now, thanks for making those changes.

**Pull request merged**
